alf/algorithms/diayn_algorithm.py

```diff
@@ -108,37 +119,40 @@ def train_step(self,
         observation, skill = observations_aug
         prev_skill = state
 
+        # normalize observation for easier prediction
+        if self._observation_normalizer is not None:
+            observation = self._observation_normalizer.normalize(observation)
+
         if self._encoding_net is not None:
-            feature, _ = self._encoding_net(observation)
+            feature = self._encoding_net(observation)
 
-        skill_pred, _ = self._discriminator_net(inputs=feature)
+        skill_pred = self._discriminator_net(inputs=feature)
 
-        skill_discriminate_loss = tf.nn.softmax_cross_entropy_with_logits(
-            labels=prev_skill, logits=skill_pred)
+        if self._skill_spec.is_discrete:
+            loss = torch.nn.CrossEntropyLoss(reduction='none')(
+                input=skill_pred, target=prev_skill.to(torch.int64))
+        else:
+            loss = torch.nn.MSELoss(reduction='none')(
+                input=skill_pred, target=prev_skill)
 
-        valid_masks = tf.cast(
-            tf.not_equal(step_type, StepType.FIRST), tf.float32)
-        skill_discriminate_loss = skill_discriminate_loss * valid_masks
+        valid_masks = (step_type != StepType.FIRST).to(torch.float32)
+        loss *= valid_masks
 
         intrinsic_reward = ()
 
         if calc_intrinsic_reward:
-            # use negative cross-entropy as reward
-            # neglect neg-prior term as it is constant
-            intrinsic_reward = tf.stop_gradient(-skill_discriminate_loss)
+            intrinsic_reward = -loss.detach()
             intrinsic_reward = self._reward_normalizer.normalize(
                 intrinsic_reward)
 
-        return AlgorithmStep(
-            outputs=(),
+        return AlgStep(
+            output=(),
             state=skill,
             info=DIAYNInfo(
                 reward=intrinsic_reward,
                 loss=LossInfo(
-                    loss=skill_discriminate_loss,
-                    extra=dict(
-                        skill_discriminate_loss=skill_discriminate_loss))))
+                    loss=loss, extra=dict(skill_discriminate_loss=loss))))
```

Comment:
can be optimized a little bit:
```python
return AlgStep(
    output=(),
    state=skill,
    info=DIAYNInfo(
        reward=intrinsic_reward,
        loss=loss))
```
And multiplying with `valid_masks` and make the LossInfo as `calc_loss()`.
This can reduce the number of operations because of batching. And it also reduces memory of the `extra` field.

Reply:
I cannot get StepType in calc_loss() for computing valid_masks?